### PR TITLE
inproved_battle_system

### LINKED ISSUE #30: Improvement on the Battle system, iteration 2
As a second part of our implementation of our battle system, we want to make the battle more unpredictable, while at the same time introducing the possibility to have arbitrary combatants against a single main character.
### ACCEPTANCE CRITERIA:

- Improve on the already existing battling system, by having the function accept two sets of arguments. The first will accept a pointer to a main character object, whereas the second set will be able to accept any number of creature pointers, that will act as the opposing party. The function will then have the main character contest each one of the foes in succession
- The aforementioned contest will consist of a d20 roll for each opposing side (a creature and the main character) to which the strength of each will be added. The highest total score (1d20 + strength) will be victorious (ties are in favor of the main character)
- At the end of the battle, a total score of how many victories and how many losses will be printed to the console in a presentable way
- Finally, the battle will be declared as 'won' if the total losses of the main character is less than their health. Otherwise, the battle is declared as 'lost'

### PR DESCRIPTION
**Pull request by DIONYSUS-BASOUNAS**
this is the second face of our battle system, we added the ability to roll in battle and through this to determine the outcome,
we also added the ability to fight more than one foe

closes #30

### CODE REVIEW

File: src/battle/battle.cpp [hunk 1]
```diff
@@ -3,24 +3,47 @@
...
-	if (goblin.get_hp() >= main_character.get_hp())
+	main_character->PrintStats();
+	for (const auto& foe : foes)
 	{
-		std::cout
-			<< "The Goblin has more HP than the main character!, Goblin wins!"
-			<< std::endl;
+		std::cout << "Stats for " << foe->get_creature_type() << ":" << std::endl;
+		foe->PrintStats();
+		int foe_roll = utils::RollDice(1, 20) + foe->get_strength();
```
Comment (achilleasmitos) [inline src/battle/battle.cpp:18]: new line before int declaration

File: src/creature/creature.h [hunk 2]
```diff
@@ -21,8 +21,9 @@ class Creature
 		std::string creature_type);
 
 	void PrintStats();
-
 	int get_hp();
+	std::string get_creature_type();
+	int get_strength();
```
Comment (achilleasmitos) [inline src/creature/creature.h:26]: this getter is super-specific. would you mind having a getter for all of the ability scores, if possible?

File: src/main.cpp [hunk 3]
```diff
@@ -57,7 +57,21 @@ int main()
 		std::cout << "Rolling with advantage: " << utils::RollWithAdvantage() << std::endl;
 		std::cout << "Rolling with disadvantage: " << utils::RollWithDisadvantage()
 				  << std::endl;
-		Battle();
+		Creature goblin(11, 14, 35, 16, 14, 10, 8, 7, 12, "Goblin");
```
Comment (achilleasmitos) [inline src/main.cpp:60]: I believe it's time we start allocating the different creatures in the heap memory. Please turn this into a heap allocation, and treat it accordingly

**Review by achilleasmitos — CHANGES_REQUESTED** [review]
Nice attempt. Apart from a couple of things here and there, I believe that it is well-made!

File: src/creature/creature.h (hunk 2 above)
Comment (DIONYSUS-BASOUNAS) [reply src/creature/creature.h:26]: have you found a better solution for this? it's probably very easy but i just can't make it work

Comment (achilleasmitos) [reply src/creature/creature.h:26]: Perhaps this answer from this post will help you:
https://stackoverflow.com/questions/3473438/return-array-in-a-function

Basically, treat the ability_scores array like a pointer to its first element. You can return (from the getter function) that pointer, and store it in an appropriate value outside the getter (say, mc_ability_scores). You then access any value you like using normal array syntax, ege mc_ability_scores[0] should get you the strength score. Give it a try :)

Comment (achilleasmitos) [reply src/creature/creature.h:26]: On a second thought, perhaps this isn't the best idea, exposing the address of a protected member like that. I believe having a getter function for all ability scores that accepts an integer parameter (as large as the number of ability scores) then return the value of that particular ability score, is our best option.

So something along the lines of:
`int get_ability_score(int n)
{
if (n > -1 && n < 6)
{
return m_ability_scores[n];
}
}`

File: src/creature/creature.cpp [hunk 4]
```diff
@@ -39,4 +40,17 @@ void Creature::PrintStats()
 	std::cout << "==============================" << std::endl;
 }
 
-int Creature::get_hp() { return m_hp; }
+int Creature::get_hp() { return m_hp; };
+int Creature::get_ability_score(int n)
+{
+	if (n >= -1 && n < 6)
```
Comment (achilleasmitos) [inline src/creature/creature.cpp:46]: n must be 0 or greater. so simply `n > -1` here

File: src/creature/creature.cpp [hunk 5]
```diff
@@ -39,4 +40,17 @@ void Creature::PrintStats()
...
+	{
+		return m_ability_scores[n];
+	}
+	else
+	{
+		// Return a default value for invalid ability scores
```
Comment (achilleasmitos) [inline src/creature/creature.cpp:52]: I like the negative return value here. I would also have the following statement before the return:
`std::cerr << "You have asked for an invalid ability score. Valid values are between 0 and 6." << std::endl;`

File: src/creature/creature.cpp [hunk 6]
```diff
@@ -39,4 +40,17 @@ void Creature::PrintStats()
...
+		return -1;
+	}
+};
+std::string Creature::get_creature_type() { return m_creature_type; };
```
Comment (achilleasmitos) [inline src/creature/creature.cpp:56]: insert empty line above

**Review by achilleasmitos — COMMENTED** [review]
It's coming along nicely. Just a bit more tidying-up to do.

File: src/creature/creature.cpp [hunk 7]
```diff
@@ -39,4 +40,20 @@ void Creature::PrintStats()
 	std::cout << "==============================" << std::endl;
 }
 
-int Creature::get_hp() { return m_hp; }
+int Creature::get_hp() { return m_hp; };
```
Comment (achilleasmitos) [inline src/creature/creature.cpp:43]: FYI, it's customary to not have any `;` after function definitions (see `PrintStats` above). Also, we separate function definitions by an empty line.

**Review by achilleasmitos — APPROVED** [review]

**Pull request merged**